ctru-rs/src/services/uds.rs

```diff
@@ -0,0 +1,1273 @@
+
+/// Handle to the UDS service.
+pub struct Uds {
+    _service_handler: ServiceReference,
+    context: Option<ctru_sys::udsBindContext>,
+    network: Option<ctru_sys::udsNetworkStruct>,
+    scan_buf: Box<[u8; Self::SCAN_BUF_SIZE]>,
+}
+
+static UDS_ACTIVE: Mutex<()> = Mutex::new(());
+
+impl Uds {
+    /// Size of one frame.
+    const RECV_FRAME_SIZE: usize = ctru_sys::UDS_DATAFRAME_MAXSIZE as usize;
+
+    /// Size of receive buffer; max frame size * 8.
+    const RECV_BUF_SIZE: u32 = ctru_sys::UDS_DEFAULT_RECVBUFSIZE;
+
+    /// Shared memory size; must be slightly larger
+    /// than `RECV_BUF_SIZE`.
+    const SHAREDMEM_SIZE: usize = 0x3000;
+
+    /// Buffer used while scanning for networks.
+    /// This value is taken from the devkitPRO example.
+    const SCAN_BUF_SIZE: usize = 0x4000;
+
+    /// The maximum number of nodes that can ever be connected
+    /// to a network (16). Can be further limited.
+    const MAX_NODES: u8 = ctru_sys::UDS_MAXNODES as u8;
+
+    /// The maximum amount of app data any server can provide.
+    /// Limited by the size of a struct in libctru.
+    const MAX_APPDATA_SIZE: usize = 200;
+
+    /// Retrieve the current status of the service.
+    pub fn service_status(&self) -> ServiceStatus {
+        match (self.context, self.network) {
+            (None, None) => ServiceStatus::Disconnected,
+            (Some(_), None) => ServiceStatus::Client,
+            (Some(_), Some(_)) => ServiceStatus::Server,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Initialise a new service handle.
+    /// No `new_with_buffer_size` function is provided, as there isn't really a
+    /// reason to use any size other than the default.
+    ///
+    /// The `username` parameter should be a max 10-byte (not 10 code point!) UTF-8 string, converted to UTF-16 internally.
+    /// Pass `None` to use the 3DS's configured username.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the [`Uds`] service is already being used,
+    /// or if the provided username is invalid (longer than 10 bytes or contains a NULL byte).
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # let _runner = test_runner::GdbRunner::default();
+    /// # use std::error::Error;
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// #
+    /// use ctru::services::uds::Uds;
+    ///
+    /// let uds = Uds::new(None)?;
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[doc(alias = "udsInit")]
+    pub fn new(username: Option<&str>) -> Result<Self, Error> {
+        if let Some(n) = username {
+            if n.len() > 10 {
+                return Err(Error::UsernameTooLong);
+            }
+        }
+        let cstr = username.map(CString::new);
+        let cstr = if let Some(conv) = cstr {
+            match conv {
+                Ok(c) => Some(c),
+                Err(_) => return Err(Error::UsernameContainsNull),
+            }
+        } else {
+            None
+        };
```

Comment:
I think you could simply this a bit if we had an `impl From<std::ffi::NulError> for Error`:

```suggestion
        let cstr = username.map(CString::new).transpose()?;
```

Reply:
I'm not happy with most of the error handling in this module, so spending some time reworking it all would be useful I think.